File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.js
/**
 * @param {number[][]} matrix
 * @return {number}
 */
var minFallingPathSum = function(matrix) {
    let rows = matrix.length;
    let cols = matrix[0].length;
    
    for(let row = 1; row < rows; row++){
        for(let col = 0; col < cols; col++){
            let prevRow = row-1;
            let curCost = matrix[row][col];
            let minCost = matrix[prevRow][col] + curCost;
            
            if(col-1>=0){
                minCost = Math.min(minCost, curCost + matrix[prevRow][col-1]);
            }
            
            if(col+1<cols){
                minCost = Math.min(minCost, curCost + matrix[prevRow][col+1]);
            }
            matrix[row][col] = minCost;
        }    
    }
    
    let minPathSum = Infinity;
    
    for(let i=0;i<cols;i++){
        minPathSum = Math.min(minPathSum, matrix[rows-1][i]);
    }
    return minPathSum;
};

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int rows=matrix.length;
        int cols=matrix[0].length;
        
        int[][] dp = new int[rows][cols];
        
        // fill first row
        for(int i=0;i<cols;i++){
            dp[0][i]=matrix[0][i];
        }
        
        // fill other rows
        
        for(int i=1;i<rows;i++){
            for(int j=0;j<cols;j++){
                int cur_value = matrix[i][j];
                int prev_best = dp[i-1][j];
                if(j-1>=0){
                    prev_best = Math.min(prev_best,dp[i-1][j-1]);
                }
                if(j+1<cols){
                    prev_best = Math.min(prev_best,dp[i-1][j+1]);
                }
                dp[i][j] = cur_value + prev_best;
            }
        }
        
        int lowest_sum = Integer.MAX_VALUE;
        
        for(int i=0;i<cols;i++){
            lowest_sum = Math.min(lowest_sum,dp[rows-1][i]);
        }
        return lowest_sum;
    }
}